Src/VimCore/SelectionChangeTracker.fs

```diff
@@ -34,6 +34,9 @@ type internal SelectionChangeTracker
     /// key input and not in Visual Mode 
     let mutable _selectionDirty = false
 
+    /// Whether the text view has multiple carets
+    let mutable _hasMultipleCarets = false

```

Comment:
Did you consider implementing this as an `IExternalEditAdapter` implementation? That is where my first attempt would've been to fix this. If it didn't work I'd be interested to know why.

Reply:
I did look at and try to comprehend the external edit adapter/monitor infrastructure. In my admittedly cursory review, I concluded that it was designed to use events for triggering entry to the mode. Furthermore, that core concept being probed was a specific kind of tag.

In the case of multiple carets, we are hinging off of selection changes. But because VsVim also reacts to external selection changes, we would have an ordering dilemma because it is critical that the selection changed event not be allowed to propagate to the selection change tracker otherwise it will switch to some visual or selection mode, clobber the selection and remove the multiple carets. As a result, I concluded that the selection change tracker needs to be aware of multiple carets and incorporate their presence into its overall (very careful) selection change handling.

This is not to say that the external edit adapter API could not be extended to handle this use case but that out-of-the-box it did not seem like good fit. However, if you now know that we need to react to selection changed events, and probably need the cooperation of the selection changed handler to prevent its interference, you might be able to imagine an approach that would be more general, less invasive, more encapsulated, etc.

Reply:
I should clean up the API so it's clearer that it's not purely tag based. It definitely can be implemented on the basis of tags but that's optional. It can also be implemented with a simple on / off type of implementation (the later ReSharper interfaces are done that way). 

But I agree the ordering is problematic here based on the current implementation. I think I could refactor this a bit too work but that may be a bit overkill. The main purpose of that interface was to allow external components to easily plug into this system. This is pretty core so makes sense that it should go directly in this case.